src/openms/include/OpenMS/METADATA/ID/MoleculeQueryMatch.h

```diff
@@ -0,0 +1,171 @@
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Hendrik Weisser $
+// $Authors: Hendrik Weisser $
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_METADATA_ID_MOLECULEQUERYMATCH_H
+#define OPENMS_METADATA_ID_MOLECULEQUERYMATCH_H
+
+#include <OpenMS/METADATA/ID/MetaData.h>
+#include <OpenMS/METADATA/ID/IdentifiedCompound.h>
+#include <OpenMS/METADATA/ID/IdentifiedSequence.h>
+#include <OpenMS/METADATA/PeptideHit.h> // for "PeakAnnotation"
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/composite_key.hpp>
+#include <boost/variant.hpp>
+
+namespace OpenMS
+{
+  namespace IdentificationDataInternal
+  {
+    // @TODO: move "PeakAnnotation" out of "PeptideHit"
+    typedef std::vector<PeptideHit::PeakAnnotation> PeakAnnotations;
+    typedef std::map<ProcessingStepRef, PeakAnnotations> PeakAnnotationSteps;
+
+    typedef boost::variant<IdentifiedPeptideRef, IdentifiedCompoundRef,
+                           IdentifiedOligoRef> IdentifiedMoleculeRef;
+
+    /*!
+      Meta data for a search hit (e.g. peptide-spectrum match).
+    */
+    struct MoleculeQueryMatch: public ScoredProcessingResult
+    {
+      IdentifiedMoleculeRef identified_molecule_ref;
+
+      DataQueryRef data_query_ref;
+
+      Int charge;
+
+      // peak annotations (fragment ion matches), potentially from different
+      // data processing steps:
+      PeakAnnotationSteps peak_annotations;
+
+      explicit MoleculeQueryMatch(
+        IdentifiedMoleculeRef identified_molecule_ref,
+        DataQueryRef data_query_ref, Int charge = 0,
+        const ScoreList& scores = ScoreList(),
+        const std::vector<ProcessingStepRef>& processing_step_refs =
+        std::vector<ProcessingStepRef>(),
+        const PeakAnnotationSteps& peak_annotations = PeakAnnotationSteps()):
+        ScoredProcessingResult(scores, processing_step_refs),
+        identified_molecule_ref(identified_molecule_ref),
```

Comment:
this is a problem, you shadow your member parameter here! you should probably use 
```
IdentifiedMoleculeRef identified_molecule_ref_tmp,
...
identified_molecule_ref(identified_molecule_ref_tmp)
```

Reply:
I disagree. See here: https://github.com/OpenMS/OpenMS/pull/3746#issuecomment-427810167